openapi.yaml

```
@@ -4992,36 +5100,55 @@ components:
           nullable: true
       allOf:
         - $ref: '#/components/schemas/process'
-    process_namespace:
+    process_graph_namespace:
       type: string
       nullable: true
       default: null
       example: null
       description: |-
         The namespace the `process_id` is valid for.
 
-        The following options are pre-defined by the openEO API, but additional
+        The following options are predefined by the openEO API, but additional
         namespaces may be introduced by back-ends or in a future version of the API.
         
-        * `null` (default): Checks both user-defined and pre-defined processes,
+        * `null` (default): Checks both user-defined and predefined processes,
            but prefers user-defined processes if both are available.
-           This allows users to add missing pre-defined processes for portability,
+           This allows users to add missing predefined processes for portability,
            e.g. common processes from [processes.openeo.org](https://processes.openeo.org)
            that have a process graph included.
            It is RECOMMENDED to log the namespace selected by the back-end for debugging purposes.
-        * `backend`: Uses exclusively the pre-defined processes listed at `GET /processes`.
-        * `user`: Uses exclusively the user-defined processes listed at `GET /process_graphs`.
+        * `backend`: Uses exclusively the predefined processes listed at `GET /processes`.
+        * `user`: Uses exclusively the user-defined processes listed at `GET /processes/@{user_id}`
+          with `{user_id}` being the user id of the authenticated user. This is an alias for 
+          directly specifying `@{user_id}` as namespace.
 
         If multiple processes with the same identifier exist, Clients SHOULD
         inform the user that it's recommended to select a namespace.
+    process_namespace:
+      type: string
```

I'm struggling with that the namespace allows so many characters. It's hard then to convert those into code and back, e.g. in the Web Editor Math Editor. 

```suggestion
      type: string
      pattern: ^@?[\w\-\.~]+$
```

---

FYI: we started playing with URL based namespaces to allow UDP sharing, as documented here: https://open-eo.github.io/openeo-python-client/cookbook/udp_sharing.html#using-a-public-udp-through-url-based-namespace
To allow that, the namespace pattern has to be more liberal I'm afraid

---

Then I won't be able to handle them in JS, I'm afraid. What do you need the other characters for? Other characters are a pain to be used in (generated) code. I was actually only implementing this for a request from VITO (see https://github.com/Open-EO/openeo-web-editor/issues/178) so one of the requests will then not be solvable.

---

I'm not sure we are talking about the same thing or at least I don't completely understand your struggle with namespace characters.
what I am referring to is using a URL as namespace in the JSON representation of a process graph. Like in the example I linked to:
```json
    "to_celsius": {
        "process_id": "fahrenheit_to_celsius",
        "namespace": "https://openeo.vito.be/openeo/1.0/processes/u:johndoe/fahrenheit_to_celsius",
        "arguments": {"fahrenheit": 86}
    }
```

In my current understanding a namespace is just an opaque string, taking one of the predefined values (like `backend` or `user`), or some backend-specific  value/format. I don't think the namespace has to be parsed or encoded/decoded at the level of the web editor (e.g. in Python client you just pass a string to a `namespace` argument, and the client is not aware what namespaces are valid or what their format should be)

---

This is a misunderstanding, indeed. This restriction is not in the process graph nodes, but in the URL of the /processes/{namespace}/... endpoints, where {namespace} can only be the given characters.

The process graph node namepsaces can still include arbritrary URLs (although not supported by the JS ecosystem and R has no namespace support at all right now).

---

By the way, why not simply...

```json
    "to_celsius": {
        "process_id": "fahrenheit_to_celsius",
        "namespace": "u:johndoe", // or @johndoe
        "arguments": {"fahrenheit": 86}
    }
```
?

---

> but in the URL of the /processes/{namespace}/... endpoints, where {namespace} can only be the given characters.

ok then indeed a restricted character set makes sense. I'm fine with alphanumeric and a couple of special characters that can act as some kind of separator. I don't have strong preference.